Prevent ripple from overflowing Android picker

redstar504 opened the pull request:
@sobitneupane @chiragsalian

Fixes regression of https://github.com/Expensify/App/pull/16103

### Details

After some further review, it looks like we can combine both proposed solutions for fixing the ripple overflow. For older Android, we can inset the items, and for >= v31 we can use the `android:clipToOutline` property for a cleaner look.

This is achieved by using the version specific values directories that Android supports.  In the screenshots area below I've provided a recording with side-by-side comparison for devices running Android v29 and Android v33.

This is optional of course, if we want it to look identical on old and new we can drop the second commit.

### Fixed Issues
<!---
1. Please replace GH_LINK with a URL link to the GitHub issue this Pull Request is fixing.
2. Please replace PROPOSAL: GH_LINK_ISSUE(COMMENT) with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>
$ https://github.com/Expensify/App/issues/<number-of-the-issue(comment)>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/15460
PROPOSAL: https://github.com/Expensify/App/issues/15460#issuecomment-1493273780


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Sign into the Android native app
2. Navigate to a page containing a picker
3. Open the picker
4. Scroll to a position where the top item is partially out of the scroll view of the picker
5. Long press on the item and confirm that the ripple effect does not extend beyond the background of the picker
6. Scroll to the end of the list, to a position where the bottom item is partially out of view
7. Long press on the item and confirm that the ripple does not extend beyond the background of the picker

On Android 11 and below, the ripple effect should be clipped at the "inset", and on on Android 12 and above, the ripple effect should be clipped at the outline.  See Android testing videos for proper display expectations.

Example pages containing a picker:
* Workspaces -> Workspace -> General Settings -> Default Currency
* Workspaces -> Workspace -> Reimburse Expenses -> Track Distance (Unit)
* Settings -> Profile -> Personal Details -> Home Address -> Country



- [X] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

1. Sign into the Android native app
2. Navigate to a page containing a picker
3. Open the picker
4. Scroll to a position where the top item is partially out of the scroll view of the picker
5. Long press on the item and confirm that the ripple effect does not extend beyond the background of the picker
6. Scroll to the end of the list, to a position where the bottom item is partially out of view
7. Long press on the item and confirm that the ripple does not extend beyond the background of the picker

On Android 11 and below, the ripple effect should be clipped at the "inset", and on on Android 12 and above, the ripple effect should be clipped at the outline.  See Android testing videos for proper display expectations.

Example pages containing a picker:
* Workspaces -> Workspace -> General Settings -> Default Currency
* Workspaces -> Workspace -> Reimburse Expenses -> Track Distance (Unit)
* Settings -> Profile -> Personal Details -> Home Address -> Country



### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
9. Upload an image via copy paste
10. Verify a modal appears displaying a preview of that image
--->

1. Sign into the Android native app on the staging environment
2. Navigate to a page containing a picker
3. Open the picker
4. Scroll to a position where the top item is partially out of the scroll view of the picker
5. Long press on the item and confirm that the ripple effect does not extend beyond the background of the picker
6. Scroll to the end of the list, to a position where the bottom item is partially out of view
7. Long press on the item and confirm that the ripple does not extend beyond the background of the picker

On Android 11 and below, the ripple effect should be clipped at the "inset", and on on Android 12 and above, the ripple effect should be clipped at the outline.  See Android testing videos for proper display expectations.

Example pages containing a picker:
* Workspaces -> Workspace -> General Settings -> Default Currency
* Workspaces -> Workspace -> Reimburse Expenses -> Track Distance (Unit)
* Settings -> Profile -> Personal Details -> Home Address -> Country



- [X] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [X] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

[Screencast from 2023-03-19 11:30:39 AM.webm](https://user-images.githubusercontent.com/1311325/226199050-33f867cd-10af-4756-9104-3a183bb6f294.webm)

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/1311325/226198823-2f7f3c70-5604-497d-b928-0d49e8fafb63.mp4

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/1311325/226198838-4a2491db-5dcd-4171-be25-30b30541ab80.mp4

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/1311325/226198847-216ca91a-310e-458f-9cba-dfd1dadb9ddd.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/1311325/226198858-9a3ecdad-2302-4dd8-8063-399d151a9cb6.mov

</details>

<details>
<summary>Android</summary>

####
__Left:__ Inset method (< API 31) | __Right:__ ClipToOutline method (>= API 31)

[Screencast from 2023-04-05 21-03-10.webm](https://user-images.githubusercontent.com/1311325/230269306-feb88eaa-0c75-46e1-9b8c-67d177f5b275.webm)

![Screenshot from 2023-04-05 20-59-55](https://user-images.githubusercontent.com/1311325/230268809-bcdc1933-b02d-44a1-a6cc-95792397c100.png)

</details>

redstar504 committed:
Prevent ripple from overflowing drawable

redstar504 committed:
Use version specific styles

sobitneupane commented:
### Screenshots/Videos

API 29

https://user-images.githubusercontent.com/25876548/230377569-92311be4-8e1b-4fd9-9537-aacce1a6c77d.mov

API 32

https://user-images.githubusercontent.com/25876548/230377615-2993472b-2912-4bd8-99a9-87e379a047ff.mov

sobitneupane commented:
@redstar504 The branch has merge conflict. Can you please resolve it and let me know.

redstar504 committed:
Merge branch 'main' into Brayden-AndroidPickerFix

redstar504 committed:
Remove unnecessary style statement

redstar504 commented:
@sobitneupane Conflicts resolved.  There was also one style declaration I added that is not necessary so I removed it.  Tested again on API 29 and 33 and works as expected.

redstar504 commented on android/app/src/main/res/values-v31/styles.xml:
```diff
@@ -1,8 +1,5 @@
 <resources>
 
-    <!-- App Theme for Android versions 12 and above -->
-    <style name="AppTheme" parent="BaseAppTheme"/>
```
App theme will be inherited from the one defined in `values-v26/styles.xml`, so not necessary.

sobitneupane reviewed (commented):
@redstar504 Can you please add Test Steps, Offline Test Steps and QA steps.

sobitneupane reviewed (commented):
### Screenshots/Videos

<details>
<summary>Android</summary>

AIP<31

https://user-images.githubusercontent.com/25876548/230597154-b0b649c4-d8c0-4880-bb01-81aefe5ec590.mov

API>=31

https://user-images.githubusercontent.com/25876548/230597841-b84890cb-56cc-4efc-b524-bfd2ec6f7d23.mov

</details>

<details>
<summary>Web</summary>
<img width="331" alt="Screenshot 2023-04-07 at 16 36 33" src="https://user-images.githubusercontent.com/25876548/230597069-a36ab308-84dc-40c2-94f7-36154828ef12.png">


</details>

<details>
<summary>Mobile Web - Chrome</summary>
<img src="https://user-images.githubusercontent.com/25876548/230597221-10124722-c91a-43f4-a7cb-a51668923f3b.png" width="350">


</details>

<details>
<summary>Mobile Web - Safari</summary>
<img src="https://user-images.githubusercontent.com/25876548/230597208-dffa1e68-fb50-4128-8496-cf5358d19952.png" width="350">

</details>

<details>
<summary>Desktop</summary>
<img width="401" alt="Screenshot 2023-04-07 at 16 44 50" src="https://user-images.githubusercontent.com/25876548/230597827-79cff42c-7d2b-45ad-a6dd-0bad108e58d4.png">


</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/25876548/230597114-f3c984d5-0178-4f4f-9726-70fd61a2a63b.mov

</details>

redstar504 commented:
@sobitneupane I have added the test steps.

sobitneupane reviewed (approved):
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/4671952159) to staging by https://github.com/chiragsalian in version: 1.2.99-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|failure ❌ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/4684921351) to production by https://github.com/francoisl in version: 1.2.99-6 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
